### PR TITLE
introduce useFacet

### PR DESCRIPTION
**Pull request by andrzejewsky**

### CODE REVIEW

**Comment by andrzejewsky** [comment]
TODO:
- write tests
- update docs

File: packages/core/nuxt-theme-module/theme/helpers/ui/index.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
+
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/helpers/ui/index.ts:1]: these fns definitely need descriptions on how to be implemented and what they do (ideally a document in docs so warns can just link to it) and should be in the boilerplate but lets leave it for our discussion

File: packages/core/nuxt-theme-module/theme/pages/Category.vue [hunk 2]
```diff
@@ -250,111 +249,51 @@ import {
...
-  { text: 'Women', route: { link: '#' } }
-];
-
 export default {
   transition: 'fade',
   setup(props, context) {
-    const { query } = context.root.$route;
+    const th = createThemeHelpers();
     onMounted(() => context.root.$scrollTo(context.root.$el, 2000));
+    const isGridView = ref(true);
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/pages/Category.vue:263]: how about making an object called `ui` for such controls

File: packages/core/nuxt-theme-module/theme/pages/Category.vue [hunk 3]
```diff
@@ -250,111 +249,51 @@ import {
...
-    const filters = ref(null);
-    const loading = computed(() => categoriesLoading.value || productsLoading.value);
-    const productsSearchParams = computed(() => ({
-      catId: (categories.value[0] || {}).id,
-      page: currentPage.value,
-      perPage: itemsPerPage.value,
-      filters: filters.value,
-      sort: sortBy.value
-    }));
+    const products = computed(() => facetGetters.getResults(searchData.value));
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/pages/Category.vue:270]: if thats for products why we dont call this getter getProducts ?

File: packages/core/nuxt-theme-module/theme/pages/Category.vue [hunk 4]
```diff
@@ -250,111 +249,51 @@ import {
...
+    const categoryTree = computed(() => facetGetters.getCategoryTree(searchData.value));
+    const breadcrumbs = computed(() => facetGetters.getBreadcrumbs(searchData.value));
+    const sortBy = computed(() => facetGetters.getSortOptions(searchData.value));
+    const facets = computed(() => facetGetters.getGroupedFacets(searchData.value, ['color', 'size']));
+    const pagination = computed(() => facetGetters.getPaginationInfo(searchData.value));
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/pages/Category.vue:275]: why not just `getPagination`

File: packages/core/nuxt-theme-module/theme/pages/Category.vue [hunk 5]
```diff
@@ -250,111 +249,51 @@ import {
...
 
     onSSR(async () => {
-      await search(getCategorySearchParameters(context));
-      await productsSearch(productsSearchParams.value);
-      filters.value = getFiltersFromUrl(context, availableFilters.value);
-      await productsSearch(productsSearchParams.value);
+      await search(th.getFacets());
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/pages/Category.vue:278]: (theme helpers are completely different discussion but just leaving it here so i wont forget)

`getFacets()` is not descriptive at all, is it `getFacetsFromURL` or what? why this method has the same name as facet getter and why its not aprametrized (pure) ;d

File: packages/commercetools/composables/src/getters/categoryGetters.ts [hunk 6]
```diff
@@ -6,6 +6,8 @@ export const getCategoryTree = (category: Category): AgnosticCategoryTree | null
   const buildTree = (rootCategory: Category) => ({
     label: rootCategory.name,
     slug: rootCategory.slug,
+    id: rootCategory.id,
+    isSelected: rootCategory.id === category.id,
```
[the remark this answers is not in the document]
Comment (andrzejewsky) [reply packages/commercetools/composables/src/getters/categoryGetters.ts:10]: rename to isCurrent

File: packages/commercetools/composables/src/getters/_utils.ts [hunk 7]
```diff
@@ -102,3 +103,63 @@ export const createFormatPrice = (price: number) => {
...
+
+export const buildBreadcrumbs = (rootCat) =>
+  buildBreadcrumbsList(rootCat, [])
+    .reverse()
+    .reduce((prev, curr, index) => ([
+      ...prev,
+      { ...curr, link: `${prev[index - 1]?.link || '' }/${curr.link}` }]),
+    []);
+
+const filterFacets = criteria => f => criteria ? criteria.includes(f) : true;
```
[the remark this answers is not in the document]
Comment (andrzejewsky) [reply packages/commercetools/composables/src/getters/_utils.ts:120]: split to files

File: packages/commercetools/composables/src/types/index.ts [hunk 8]
```diff
@@ -17,3 +18,23 @@ export interface ProductsSearchParams {
   slug?: string;
   id?: string;
 }
+
+export interface FacetSearchInput {
```
[the remark this answers is not in the document]
Comment (andrzejewsky) [reply packages/commercetools/composables/src/types/index.ts:22]: NOTE: generic interface for input for the most common cases { catId, filters, sort, page ... [x: string]: any }

File: packages/commercetools/composables/src/useFacet/index.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import { useFacetFactory, FacetSearchData } from '@vue-storefront/core';
+import { getProduct, getCategory, AttributeType } from '@vue-storefront/commercetools-api';
+import { enhanceProduct, getFiltersFromProductsAttributes } from './../helpers/internals';
+import { ProductVariant } from './../types/GraphQL';
+import { FacetSearchInput, FacetResultsData } from './../types';
+
+const factoryParams = {
+  search: async (params: FacetSearchData<FacetResultsData, FacetSearchInput>): Promise<FacetResultsData> => {
+    const perPageOptions = [20, 40, 100];
```
[the remark this answers is not in the document]
Comment (andrzejewsky) [reply packages/commercetools/composables/src/useFacet/index.ts:9]: NOTE: move to the config file and read from api-client

File: packages/core/nuxt-theme-module/theme/components/Filters.vue [hunk 10]
```diff
@@ -1,6 +1,139 @@
...
+      console.log(props.facets);
+      selectedFilters.value = props.facets.reduce((prev, curr) => ({
+        ...prev,
+        [curr.id]: curr.options
+          .filter(o => o.selected)
+          .map(o => o.id)
+      }), {});
+    });
+
+    const isSelected = (facet, option) => (selectedFilters.value[facet.id] || []).includes(option.id);
```
[the remark this answers is not in the document]
Comment (andrzejewsky) [reply packages/core/nuxt-theme-module/theme/components/Filters.vue:105]: NOTE: isFilterSelected

File: packages/core/nuxt-theme-module/theme/helpers/ui/index.ts (hunk 1 above)
Comment (andrzejewsky) [reply packages/core/nuxt-theme-module/theme/helpers/ui/index.ts:1]: NOTE: replace factory to object

File: packages/core/nuxt-theme-module/theme/pages/Category.vue [hunk 11]
```diff
@@ -250,115 +249,48 @@ import {
...
-    const { categories, search, loading: categoriesLoading } = useCategory('categories');
-    const {
-      products: categoryProducts,
-      totalProducts,
-      search: productsSearch,
-      loading: productsLoading,
-      availableFilters,
-      availableSortingOptions
-    } = useProduct('categoryProducts');
+    const th = createThemeHelpers();
```
Comment (filrak) [inline packages/core/nuxt-theme-module/theme/pages/Category.vue:263]: useUiHelpers()
useUiState() for `ui-state.ts`

**Review by filrak — APPROVED** [review]

**Pull request merged**